_posts/2025-03-18-collecting-diagnostic-data.md

```diff
@@ -0,0 +1,112 @@
+Without access to the right diagnostic data, identifying the root cause of an issue can become a time-consuming process.
+Strimzi, the Kubernetes-native Kafka operator, provides a [bash script](https://github.com/strimzi/strimzi-kafka-operator/blob/0.45.0/tools/report.sh) to collect cluster artifacts, including logs and configurations.
+
+Sometimes, this is not enough and we need long-term storage to collect custom and advanced diagnostic data such as heap dumps and flame graphs.
+Fortunately, Strimzi recently introduced the **additional volumes** feature that makes this task really easy.
+
+<!--more-->
+
+### The need for diagnostic data in Kafka troubleshooting
+
+Kafka clusters can encounter a range of issues, from resource exhaustion to unexpected application crashes.
+When these issues arise, collecting relevant diagnostic data such as logs, metrics, heap dumps, and thread dumps is essential for effective debugging.
+Without this information, developers are left guessing, often leading to unnecessary downtime and prolonged resolutions.
+
+Common diagnostic artifacts needed for debugging Kafka and Java application in general include:
+
+- **Heap dumps**: To analyze memory usage and detect memory leaks.
+- **Thread dumps**: To inspect thread states and potential deadlocks.
+- **JVM metrics**: To monitor CPU, memory, and GC activity.
+- **Log files**: To track application errors and warnings.
+- **Flame graphs**: To visually identify performance bottlenecks.
+
+Strimzi makes it easier to capture these artifacts by leveraging Kubernetes-native mechanisms such as persistent volumes.
+
+### Introducing Strimzi additional volumes
+
+The additional volumes feature, introduced via [proposal 75](https://github.com/strimzi/proposals/blob/main/075-additional-volumes-support.md), allows users to define extra storage volumes in all operands.
+This enhancement supports different types of volumes, but here we are interested in `PersistentVolumeClaims` to durably store and retrieve any kind of diagnostic data under the `/mnt` mount point.
+
+With additional volumes, users can:
+
+- Mount persistent or ephemeral storage for logs, dumps, or other debugging artifacts.
+- Capture JVM dumps without modifying the default storage configuration.
+- Enable better debugging workflows by keeping artifacts accessible even after a pod restart.
+
+### Example: capturing heap dumps with additional volumes
+
+A practical example of how additional volumes simplify debugging is demonstrated in the following procedure.
+The scenario involves collecting heap dumps from a Kafka broker when observing excessive memory consumption.
+Memory leaks can sometimes lead to out of memory exceptions (OOME) and service disruption.
+
+> [!WARNING]
+> Taking a heap dump is a heavy operation that makes the Java application hangs.
+> It is not recommended in production, unless it is not possible to reproduce the memory issue in a lower environment.
```

Comment:
@PaulRMellor I don't see the post preview. Is it because of this warning syntax?

Reply:
It is because its not 18th March yet.

Reply:
Yes. it won't show with the future date